### PR TITLE
Wip/hughsie/report per remote

### PR DESCRIPTION
**Pull request by hughsie**
This is https://github.com/hughsie/fwupd/pull/358 done perhaps a better way.

### CODE REVIEW

**Comment by superm1** [comment]
So I just a functional test with this in what should have been a failed request.  XPS 9350 that's on 1.5.1 to 1.6.1 using the fwup.efi that was missing https://github.com/rhboot/fwupdate/commit/7f0b7613c94c7a25dfffa7a1d7f1dcbd07cefc0e so it wouldn't actually try to install things.

When I rebooted fwupd thought it was a success (which I guess is expected - the UEFI plugin shows tha the last update that the BIOS actually called UpdateCapsule on was a success.

In addition to the plugin claiming that it was successful in it's last attempt I think you need one more check that shows that VersionNew that was recorded == Version.

**Comment by hughsie** [comment]
> I think you need one more check that shows that VersionNew that was recorded

Isn't that what https://github.com/hughsie/fwupd/blob/master/src/fu-engine.c#L3210 does?

**Comment by superm1** [comment]
Hum, it would appear so.  let me try again and grab the logs along the way.

**Comment by superm1** [comment]
Here's the relevant snippet:

```
19:52:19:0216 Fu  FuHistory: trying to open database '/var/lib/fwupd/pending.db'
19:52:19:0216 Fu  FuHistory: get devices
19:52:19:0216 Fu  FuHistory: got sql result 79f2996ff11608674c5d2e8db20e10e0b646f2cb
19:52:19:0216 Fu  installed version 0.1.5.1 matching history 0.1.6.1
19:52:19:0217 Fu  FuHistory: set update-state of 79f2996ff11608674c5d2e8db20e10e0b646f2cb to success


```

**Comment by hughsie** [comment]
Doh! It's the wrong way around:

	if (g_strcmp0 (fu_device_get_version (dev),
		       fwupd_release_get_version (rel_history)) != 0) {

Will fix.

**Comment by superm1** [comment]
I guess you just inverted the logic, but maybe better to use `as_utils_vercmp`  instead?

**Comment by hughsie** [comment]
indeed!

**Comment by hughsie** [comment]
Just to check as I've not got my laptop handy, does:

	if (as_utils_vercmp (fu_device_get_version (dev),
			     fwupd_release_get_version (rel_history)) >= 0) {

Work, or is that the wrong way around too?

**Comment by superm1** [comment]
```
20:01:57:0932 Fu  FuHistory: trying to open database '/var/lib/fwupd/pending.db'
20:01:57:0932 Fu  FuHistory: get devices
20:01:57:0933 Fu  FuHistory: got sql result 79f2996ff11608674c5d2e8db20e10e0b646f2cb
20:01:57:0933 Fu  performing get_results() on uefi
20:01:57:0933 Fu  FuHistory: set update-state of 79f2996ff11608674c5d2e8db20e10e0b646f2cb to success
20:01:57:0955 Fu  FuHistory: set error to 79f2996ff11608674c5d2e8db20e10e0b646f2cb: (null)
```

Looks like that still claimed success.  Actually thinking this through, you were probably on the right thought process with gstrcmp0 though.  You really want _upgrades_ **and** _downgrades_ here to report success/fail.

**Comment by hughsie** [comment]
I've polished these up and pushed to master.

**Comment by superm1** [comment]
So I 've tried with master after you did that and it's still claiming success.

That same flow is happening.  After  the reboot, it's marked to success.
```
performing get_results() on uefi
FuHistory: set update-state of 79f2996ff11608674c5d2e8db20e10e0b646f2cb to success
FuHistory: set error to 79f2996ff11608674c5d2e8db20e10e0b646f2cb: (null)
```

Doesn't there have to be an explicit failure set too if the versions aren't the same?

**Comment by hughsie** [comment]
> Doesn't there have to be an explicit failure set too if the versions aren't the same?

Maybe. I need to think a little about how to test this too. Ideas very welcome.

**Comment by hughsie** [comment]
So `get_results()` is returning success from the `uefi` plugin -- what type of failure is this exactly?

**Comment by superm1** [comment]
That one that happened in Fedora 27 for the last few months - EFI application doesn't actually call Updatecapsule()

It seems to me that there is no case that you actually want to declare success other than the versions matching.  This seems like a safe simplification:

```
        if (fu_device_get_update_state (dev) != FWUPD_UPDATE_STATE_NEEDS_REBOOT) {
                if (!fu_history_set_update_state (self->history, dev,
-                                                 fu_device_get_update_state (dev),
+                                                 FWUPD_UPDATE_STATE_FAILED,
                                                  error))
                        return FALSE;
```

**Comment by hughsie** [comment]
Isn't `fu_device_get_update_error` going to be NULL in that case? We should probably do better than just failing without setting an error. I think it might be safer to look at `fu_plugin_runner_get_results()` and if that failed, use the update-state and error-code, returning TRUE.

As a final block (the last one in the `fu_engine_update_history_device()`) we'd set the error manually to `FWUPD_UPDATE_STATE_FAILED` and then set a suitable error, although I'm not sure what that error string should be -- ideas welcome.

**Comment by superm1** [comment]
Yeah it is NULL in that case.  I'd think you can look for that scenario too.
The problem is the plugin doesn't know it failed because it's not keeping track of it's own state.  From it's perspective the ESRT will claim it was a successful update based on the last time an UEFI capsule was actually flashed.

> I think it might be safer to look at fu_plugin_runner_get_results() and if that failed, use the update-state and error-code, returning TRUE.

You mean, look at fu_device_get_update_state (dev) and if that's already marked failed, update state and error code returning true.  Otherwise set up the error.

I think this is really only a problem for the plugins that can update on reboot like UEFI.  
Others should be in some failed state immediately and your existing logic handles them.
How about this: "Failed to run update on reboot".

**Comment by hughsie** [comment]
Does https://github.com/hughsie/fwupd/pull/362 work for you? We have to be a little careful to not mark NEEDS_REBOOT transactions as FAILED when we restart the fwupd daemon.
